File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {

        int n = nums.size();

        vector<int> result(n);
        stack<int> st;

        //that for loop create a hypothetical circulation
        for (int i = 2 * n - 1; i >= 0; i--) {

            while (!st.empty() && st.top() <= nums[i % n]) {
                st.pop();
            }

            if (i < n) {
                result[i] = st.empty() ? -1 : st.top();
            }

            // Push current element
            st.push(nums[i % n]);
        }

        return result;
    }
};

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] result= new int[n];
        Arrays.fill(result, -1);
        for(int i =0; i<n;i++){
            for(int j =i+1;j<i+n; j++){
                int ind = j%n;
                if(nums[ind]>nums[i]){
                    result[i] = nums[ind];
                    break;
                }
            }
        }
        return result;
    }
}